File: 155-min-stack/min-stack.cpp
class Node{
    public:
    int val;
    Node* next;
    int minimum;
    Node(int tempData = 0, Node* tempNext = nullptr, int tempMin=INT_MAX){
        val = tempData;
        next = tempNext;
        minimum = tempMin;
    }
};

class MinStack {
    Node* head;
public:
    MinStack() {
        head = nullptr;
    }
    
    void push(int val) {
        int prevMin = head ? head->minimum : INT_MAX;
        int newMin = min(val,prevMin);
        Node* node = new Node(val, head, newMin);
        head = node;
    }
    
    void pop() {
        Node* temp = head;
        head = head->next;
        delete temp;
    }
    
    int top() {
        return head ? head->val : -1;
    }
    
    int getMin() {
        return head->minimum;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/min-stack.js

var MinStack = function() {
    this.stack = [] // { val, min }
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    let element = {
        val,
        min:val
    }

    if(this.stack.length){
        element.min = Math.min(element.min, this.stack[this.stack.length-1].min)
    }

    this.stack.push(element);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if(this.stack.length) this.stack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return (this.stack.length ? this.stack[this.stack.length-1].val : -1)
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return (this.stack.length ? this.stack[this.stack.length-1].min : -1)
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */